package/MDAnalysis/coordinates/chain.py

```diff
@@ -417,6 +420,25 @@ def _get_local_frame(self, k):
         f = k - self._start_frames[i]
         return i, f
 
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        state['ts'] = self.ts.__deepcopy__()
+        index = self.ts.frame
+        state.pop('_ChainReader__chained_trajectories_iter', None)
+        for reader in state['readers'][:self.__active_reader_index + 1]:
+            reader.rewind()
+        self.ts = state['ts']
+        return state, index
+
+
+    def __setstate__(self, args):
+        state = args[0]
+        index = args[1]
+        self.__dict__.update(state)
+        self.__chained_trajectories_iter = self._chained_iterator()
+        for i in range(index + 1):
+            self._read_next_timestep()
```

Comment:
Why not go directly to the frame?  it should support random access

Reply:
If using `self[index]`,
the `u_pickled.trajectory.next()` will end up to Timestep 0,
because iteration is based on  `__chained_trajectories_iter`, which is not updated with `self[index]`.

I am wondering if there's other methods to directly locate the new generator to that location.

Reply:
We should make this work with random access – having to iterate through the trajectory will utterly kill performance.

Reply:
Looking at https://github.com/MDAnalysis/mdanalysis/blob/8314f7d33b08bd467f1974fdc886f60b02f226c3/package/MDAnalysis/coordinates/chain.py#L566-L574 I don't see a way to restart the iterator in the middle. But before discussing this more, two questions
1. Isn't this also a problem with the other readers or do all of them implement `__iter__()` with a simple `self._read_frame(i)`?
2. How often do we pickle in the middle of a running iterator (i.e., pickle a `base.FrameIteratorAll`):
    ```python
    for ts in u.trajectory:
           pickle.dump(f, u)
    ```
    This seems a strange thing to do – we could just raise a `RuntimeError` and say that you cannot pickle an active iterator (if we can detect it being active). And the iterator rewinds after its done so pickling _outside_ of the loop just needs to rewind and make sure that calling `__iter__()` works as expected. Or am I missing something here?

The other iteration cases (slices, indices, bools) are based on `base.FrameIteratorSliced`, `base.FrameIteratorIndices` and ultimately rely on `__getitem__` access. Looking through the `__iter__` methods of `FrameIteratorSliced` https://github.com/MDAnalysis/mdanalysis/blob/8314f7d33b08bd467f1974fdc886f60b02f226c3/package/MDAnalysis/coordinates/base.py#L954-L957 and `FrameIteratorIndices` https://github.com/MDAnalysis/mdanalysis/blob/8314f7d33b08bd467f1974fdc886f60b02f226c3/package/MDAnalysis/coordinates/base.py#L1074-L1076 I am admittedly a bit puzzled why they are different: The former rewinds and directly uses `self.trajectory[i] ` whereas the latter does not rewind and uses `self.trajectory._read_frame_with_aux(frame)` – I don't remember why the difference (@jbarnoud ?)). In any case, they exist for all readers and because they rely on `__getitem__()`, they _should_ work for the ChainReader, too.

Reply:
If we decide that we have to be able to restart in the middle then we could add optional arguments to `_chained_iterator()` along the lines of
```python
def _chained_iterator(self, restart=None): 
     """Iterator that presents itself as a chained trajectory.""" 
     if restart is not None:
         # make it a FrameIteratorSliced for the special case of
         # unpickling an active iterator
         return self[restart:]

     # normal generator: iterate through all frames
     self._rewind()  # must rewind all readers 
     for i in range(self.n_frames): 
         j, f = self._get_local_frame(i) 
         self.__activate_reader(j) 
         self.ts = self.active_reader[f] 
         self.ts.frame = i 
         yield self.ts 
```
and then `__setstate__` can do something like
```python
if had_active_iterator:             # not sure how to detect
   restart = self.ts.frame + 1    # should be the last iter frame
   self.__chained_trajectories_iter = self._chained_iterator(restart=restart)
```

But to be honest, I simply can't think of code where I would want to continue this iterator in the middle. Again, maybe I am missing something here.
```

Reply:
I admit I am still confused about all these iterators. I also don't understand why we need an extra generator in the first place? Can't we just implement it directly to `ChainReader`? 

Besides, I just realize it's not just about pickling, but some inconsistency between `ChainReader` and other readers?
```
u = mda.Universe(GRO, XTC)
u.trajectory[5]
u.trajectory.next()
< Timestep 6 with unit cell dimensions [0. 0. 0. 0. 0. 0.] >

u = mda.Universe(GRO, [XTC, XTC])
u.trajectory[5]
u.trajectory.next()
< Timestep 1 with unit cell dimensions [0. 0. 0. 0. 0. 0.] >
```
because aux and generator are disconnected.

Reply:
I think this should be classified as a bug. So I submit a PR ##2815 for that, which should also solve the problem here.